### PR TITLE
ETABS_Toolkit: Set material design parameters during push

### LINKED ISSUE #378: Material Strength not set in Push
#### Description:
When pushing materials into the BHoM, the API method for setting concrete/steel/etc. strength is not called (SetO$Material), this means that the strength for any material pushed into ETABS is set as the default (for concrete, this is 4000 psi).

#### Steps to reproduce:

image

#### Test file(s):
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Issue/BHoM/ETABS_Toolkit/Issue-%23378-Material-strength?csf=1&web=1&e=aetrbA

### PR DESCRIPTION
**Pull request by JosefTaylor**
### NOTE: Depends on 

 ### Issues addressed by this PR

 Closes #378 

 

 ### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23405-set-design-parameters-during-material-push?csf=1&web=1&e=RXiDVc

 ### Changelog

 ### Additional comments

### CODE REVIEW

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by mchaf** [comment]
I looked at this today: Overall, the compressive force, modulus of elasticity and other design parameters came in without an issue. Still, there were some other things that might be worth fixing. 

First, for both steel and concrete, there are parameters that are not defined by BHoM Materials (effective yield stress, expected rupture, nonlinear material data, etc) that need to be set. Firstly, do we want want to set a value, or do we want to make it 0? If we do set a value, we should be very clear about which parameters are being set by us, and not the user. 

Secondly, within the Material Property Design Data for steel and concrete, the "grade" parameter does not take the name of the material. It should, to avoid confusion. Upon talking to Joe, it seems like this is a problem on the ETABS side to figure out and dive in to. 

I'll be happy to look at it again after we solve this, but the overall functionality of transferring compressive strength and modulus of elasticity (the two more important properties) works, which is a win :)

**Comment by JosefTaylor** [comment]
Diving into the issues Mel raised uncovered a few more - if a user pushed a material (not a bar or property containing a material), it would trigger Create, instead of Update, and the Create checked if it already existed, which is normally superfluous. Fixing this meant that properties on existing materials are overwritten, which is usually not desired. So, I changed the default push type to CreateNonExisting, so that a user could pre-load any materials they need to use before pushing dependent types, and those materials will not be changed by an unintended update.
Unfortunately, it is apparently not possible to change the 'grade' property of a material, so when adding a material, this will be shown as the default grade for the type, i.e. 'Grade 50' for steel and "f'c 4000 psi" for concrete.
Also, for steel, ETABS defines two parameters EFy and EFu, which BHoM does not define. Rather than leave these as the value set by ETABS (always from A992 Grade 50), they are set equal to Fy and Fu respectively, and the user can change them manually in ETABS if needed.

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

File: Etabs_Adapter/CRUD/Create/Material.cs [hunk 1]
```diff
@@ -143,7 +158,130 @@ private eMatType MaterialTypeToCSI(MaterialType materialType)
...
+
+        private bool SetDesignMaterial(IMaterialFragment material)
+        {
+            Engine.Base.Compute.RecordError($"Could not set ETABS design parameters for material: {material.DescriptionOrName()}. Please set them manually in ETABS, or load the material from ETABS database prior to push.");
+            return true;
+        }
+
+        /***************************************************/
+
+        private bool IsLightweight(Concrete material)
```
Comment (IsakNaslundBh) [inline Etabs_Adapter/CRUD/Create/Material.cs:272]: This is raising a warning for _all_ non-american concrete materials. Not sure it is suitable as this warning might just lead to confusion.

We might need to either define a LightWeight parameter on the Concrete material itself, or find some other solution to this.

File: Etabs_Adapter/CRUD/Create/Material.cs [hunk 2]
```diff
@@ -143,7 +158,130 @@ private eMatType MaterialTypeToCSI(MaterialType materialType)
...
+                ref finalSlope,
+                ref frictionAngle,
+                ref dilationAngle
+                );
+
+            success &= (0 == m_model.PropMaterial.SetOConcrete_1(
+                material.DescriptionOrName(),
+                material.CylinderStrength,
+                IsLightweight(material),
+                IsLightweight(material) ? 0.75 : 1.0, //Strength reduction factor for lightweight concrete
```
Comment (IsakNaslundBh) [inline Etabs_Adapter/CRUD/Create/Material.cs:201]: the 0.75 here I presume is american code? Here I also think we need to be a bit carful, and might need to have a slight think on how we set this up properly.

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
General functionality works well, but some of the changes are a bit to american code centric, and makes pushing items raise warnings that does not make sense for for example Eurocode.

Think this needs a quick discussion before being merged in.

**Comment by JosefTaylor** [comment]
@BHoMBot check required

File: Etabs_Adapter/CRUD/Create/Material.cs (hunk 2 above)
Comment (JosefTaylor) [reply Etabs_Adapter/CRUD/Create/Material.cs]: Agreed, will not set this property, and raise a note only if the concrete appears to be lightweight

File: Etabs_Adapter/CRUD/Create/Material.cs (hunk 1 above)
Comment (JosefTaylor) [reply Etabs_Adapter/CRUD/Create/Material.cs:272]: Agreed, need some more parameters on Structural materials - similar to what we've done on Timber.

Comment (JosefTaylor) [reply Etabs_Adapter/CRUD/Create/Material.cs:272]: As below, will only raise the warning on concrete that appears lightweight (either name contains LW or density is less than 2080 kg/m^3 (~130pcf)

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check ready-to-merge

**Comment by FraserGreenroyd** [comment]
Following a request by @IsakNaslundBh , I have reviewed this PR and its mergeability given the current feature freeze for the 5.3 beta release, and am happy this PR conformed to the guidelines laid out for the previous sprint and may be merged today prior to the test beta artefacts being produced for inclusion in the 5.3 beta.

**Pull request merged**